Complete refactor of the IntelliJ Ion Plugin using Kotlin and improving the syntax formatting - release V2.

belloale opened the pull request:
Description of the removals below.

* `plugin.xml` The plugin.xml is basically the manifest used by IntelliJ to hook up the plugin implementations into the IDE, this is quite important and its the central file of any IntelliJ plugin. In this case it shows as a deletion because the `plugin.xml` file is generated by Gradle in modern plugins and uses a base `plugin.xml` as sort of a template. So this will be added with all the new code.
* JFlex.jar - is used to generate all the automatically generated code under the previously called `gen` directory. No longer will be used since IntelliJ has this built in.
* build.xml - The previous ant build file, the new version will use the official IntelliJ gradle plugin.
* Everything under gen/ was generated code by using some Java code generation tools to parse the `.bnf` and `.flex` files and generate the types needed to parse Ion.
* Everything under src/ was the previous source code which will all be replaced with Kotlin.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

desaikd commented:
Hey belloale@, Thanks for creating the PR again! Current changes looks good. Although can you rename this PR and add changes as separate commits here. It would make it easy for you as well as for us to have a single PR with all the separate commits in it.

belloale commented:
Ok I can do that.

belloale commented:
@desaikd Commits have been added, take a look, let me know what you think.

desaikd commented:
I tried the tests with this PR. For me all the formatting tests failed whereas parser tests passed. Is there something that I am missing? It gives me `Assertion Error` when trying to  run [`checkDocument`](https://github.com/amzn/ion-intellij-plugin/blob/v2/src/test/kotlin/com/amazon/ion/plugin/intellij/formatting/IonFormatterTestBase.kt#L23).

belloale commented:
@desaikd I've fixed the issue you were seeing, it was due to the plugin version number not lining up, I had left it as V1 in one place and hence the IntelliJ runtime could not find the V2 plugin.

Anyways I have fixed this issue and I added a few Github Action files which will automatically test, build and upload a build artifact. We can extend this a bit later to automate the actual plugin deployment.

zslayton commented on src/main/kotlin/com/amazon/ion/plugin/intellij/IonIcons.kt:
```diff
@@ -0,0 +1,20 @@
+object IonIcons {
+    /**
+     * Alternative file type logos:
+     * - Old nucleus looking logo - ion-file-type.ping
```
Was this meant to be `.png` instead of `.ping`?

I wish GitHub would display all of the new image assets. 😞

belloale replied:
Yes I can fix that comment.

zslayton reviewed (commented):
Thanks a lot for contributing this! It looks like a big improvement overall.

belloale commented on src/main/kotlin/com/amazon/ion/plugin/intellij/Ion.bnf:
```diff
@@ -0,0 +1,45 @@
+key ::= (QUOTE KEY_NAME QUOTE) | (QQUOTE KEY_NAME QQUOTE) | (QQQUOTE KEY_NAME QQQUOTE) | KEY_NAME {methods=[getKeyName]}
+array ::= LBRACKET [elements] RBRACKET {pin=1}
+elements ::= array_element*
+expression ::= (LPAREN [expression_operator] [atoms] RPAREN)
+expression_operator ::= (annotation)* (symbol)
```
I can change the name to `sexp`. The reason why I am picking up `expression_operator` is because there are features within the Ion plugin which would be useful to parse the first operator seperately. In my mind the BNF grammar is a combination of the language spec and useful "tags" that would make language features easier to implement.

belloale commented on src/main/kotlin/com/amazon/ion/plugin/intellij/Ion.flex:
```diff
@@ -0,0 +1,309 @@
+/* Key */
+KeyCharacter = [^:\ \n\r\t\f\\\'\"]
+
+/* Operators */
```
This was existing in the [Ion flex description](https://github.com/amzn/ion-intellij-plugin/blob/master/src/com/amazon/ion/plugin/intellij/Ion.flex#L60). I did not modify this. Before I change this, I need to do more testing to make sure it doesn't cause any other regressions. Since this was existing, I recommend just leaving it how it is and then we can open an issue or future PR to fix this specifically.

belloale replied:
After reading the spec excerpt above in more detail, I think your statement regarding each of the double values (`++`, `--`, `&&`, etc) does not follow the excerpt actually, specifically this statement:

---

 **An operator** is an **unquoted** sequence of one **or more** of the following nineteen ASCII characters: !#%&*+-./;<=>?@^`|~

---

In my interpretation of this statement, this means that "An [one] operator .... is an unquoted sequence of one or more ...." of the special operator characters. So this means the parser should interpret a sequence of operator characters as a single operator, i.e. `+=`, `|=`, `&&&&`, etc are all a single operator according to that statement.

According to your statement about double values (`++`, etc.) the following example from the excerpt should have 10 symbols but it in fact has seven symbols according to the spec:

> (a==b&&c==d)     // S-expression with seven symbols

zslayton replied:
I agree with your interpretation, that's my mistake!

What drew my attention to this was the listing of "recognized" operators. Ion doesn't consider `++` to be more special than `+++` or `++++`, so having different logic for certain operators felt a bit off.

belloale commented on src/main/kotlin/com/amazon/ion/plugin/intellij/IonElements.kt:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Elements / Symbols which contain meaning in the Ion structure.
+ */
+object IonElements {
```
This file is unused, I have removed this file.

belloale commented on src/main/kotlin/com/amazon/ion/plugin/intellij/Ion.bnf:
```diff
@@ -0,0 +1,45 @@
+ionFile ::= item_*
+private item_ ::=  value
+
+object ::= (struct | array | expression)
```
Ok I mean again this is existing, I'll try to change a few of these things.

Here is the existing BNF grammar: https://github.com/amzn/ion-intellij-plugin/blob/master/src/com/amazon/ion/plugin/intellij/Ion.bnf

belloale commented on src/main/kotlin/com/amazon/ion/plugin/intellij/Ion.bnf:
```diff
@@ -0,0 +1,45 @@
+members ::= struct_pair*
+pair ::= COMMENT* (key SEPARATOR value)
+key ::= (QUOTE KEY_NAME QUOTE) | (QQUOTE KEY_NAME QQUOTE) | (QQQUOTE KEY_NAME QQQUOTE) | KEY_NAME {methods=[getKeyName]}
+array ::= LBRACKET [elements] RBRACKET {pin=1}
```
Sure, same comment as above, did not make modifications here, this is existing.

belloale commented on src/main/kotlin/com/amazon/ion/plugin/intellij/folding/IonFoldingBuilder.kt:
```diff
@@ -0,0 +1,47 @@
+    ): Array<FoldingDescriptor> =
+        PsiTreeUtil.findChildrenOfAnyType(
+            root,
+            IonObject::class.java,
+            IonArray::class.java,
+            IonExpression::class.java
```
Sure, I will look into renaming these, this was existing within the plugin.

belloale commented on src/main/kotlin/com/amazon/ion/plugin/intellij/formatting/IonLanguageCodeStyleSettingsProvider.kt:
```diff
@@ -0,0 +1,52 @@
+        """
+        // Demo text for code style.
+        'Annotation'::AnotherAnnotation::{
+            float_key: 2.55, /* This is an inline comment */,
```
Correct, I'll fix this.

belloale commented on src/main/kotlin/com/amazon/ion/plugin/intellij/formatting/IonLanguageCodeStyleSettingsProvider.kt:
```diff
@@ -0,0 +1,52 @@
+            array_key: [ "Note A", "Note B" ],
+            other_numbers: [ 0x10FBAD, 0b110101 ],
+            struct_key: {
+                integer: 25
```
I'll fix this.

belloale commented on src/main/kotlin/com/amazon/ion/plugin/intellij/formatting/blocks/IonBlockFactory.kt:
```diff
@@ -0,0 +1,25 @@
+    ): Block =
+        when (node.elementType) {
+            STRUCT -> IonStructBlock(node, formatting, options)
+            ARRAY -> IonListBlock(node, formatting, options)
```
Ok I can look into renaming this. These were existing names.

belloale commented on src/main/kotlin/com/amazon/ion/plugin/intellij/highlight/IonColorSettingsPage.kt:
```diff
@@ -0,0 +1,70 @@
+                string: "Simple struct"
+            }
+        }
+    """.trimIndent()
```
Yes, I fixed it in both places. At this point they are the same but there is no reason they should be the same, in fact they probably should be different. We could consider putting it in a shared place but it was a rather minimal issue.

belloale commented:
@zslayton  Fixed all the above comments with the exception of the comment about `KeyCharacters` in the `Ion.flex` definition. I'll take a look at that a bit later, I think since this was existing as part of the plugin, I say we proceed with this PR and take a look at modifying this later. I need to deep dive the spec a bit more and understand how we want to change that.

zslayton commented:
> @zslayton Fixed all the above comments with the exception of the comment about KeyCharacters in the Ion.flex definition. I'll take a look at that a bit later, I think since this was existing as part of the plugin, I say we proceed with this PR and take a look at modifying this later. I need to deep dive the spec a bit more and understand how we want to change that.

That works for me.

As an aside, please avoid force pushing changes to a PR that's in progress. It causes all of the code associated with our prior comments to be discarded, making it difficult to track which code was added to address concerns we've discussed.

tgregg commented on resources/plugin-changenotes.html:
```diff
@@ -0,0 +1,6 @@
+<ul>
+    2020-12-24
+    <li>
+        Complete refactor of the Amazon Ion plugin 🎉
```
Can you please add some notes about the differences between this version and the previous version?

Are there any features in the previous version that aren't in this version, or vice-versa?

belloale replied:
I can add more details here.

belloale replied:
@tgregg FYI

1. This release doesn't publish the new version to the JetBrains public repo just yet. I'll make another PR after this one to make the changes to the GitHub Actions flow to be able to publish automatically from GitHub. We need to get the publish key from IntelliJ and figure out how to safely store that on GitHub. Currently, the way you would upgrade your plugin is by downloading the build artifacts from this merge, in fact you can just download the JAR produced by the GitHub Actions flow right now and test the plugin locally. Its quite easy.
2. I'm not sure how to specifically put this into the Changelog but the changes between the old plugin and the new plugin are pretty massive, I've asked a co-worker who had the old plugin on their IntelliJ to run Code -> Reformat Code with the old plugin and uploaded both onto a Github Gist.
> [V1 Plugin Formatting](https://gist.github.com/belloale/b2fe46b84043508e8356ca85ea6f0c9d)
> [V2 Plugin Formatting](https://gist.github.com/belloale/62e9412643881e69c89a4b9d4f2c6291)

tgregg replied:
This is great, thanks! Some major improvements in there.

belloale commented:
> > @zslayton Fixed all the above comments with the exception of the comment about KeyCharacters in the Ion.flex definition. I'll take a look at that a bit later, I think since this was existing as part of the plugin, I say we proceed with this PR and take a look at modifying this later. I need to deep dive the spec a bit more and understand how we want to change that.
> 
> That works for me.
> 
> As an aside, please avoid force pushing changes to a PR that's in progress. It causes all of the code associated with our prior comments to be discarded, making it difficult to track which code was added to address concerns we've discussed.
> 
> Edit for additional clarity: ultimately we'll squash and merge all of the commits in the PR, so it's ok to add lots of small commits with description text that's not necessarily meaningful outside of the PR. The finished product will be a single commit with a description we distill from its overall contents.

Ok cool, yeah I didn't know if you guys wanted to keep the PR commit history clean with the requested changes but I can just a bunch of small commits, no worries.

tgregg reviewed (approved):
Thanks for putting this together! Let's wait for one more approval from one of the other reviewers and then merge.

belloale commented:
@zslayton @desaikd Are any of you ready to approve this PR or any more comments? Thanks!

zslayton commented on resources/plugin-changenotes.html:
```diff
@@ -1,6 +1,33 @@
+        Bad characters recovery in incorrectly formatted Ion has been improved by adding recovery statements in the BNF grammar.
+    </li>
+    <li>
+        First comment line in an IonStruct is now indented properly.
```
```suggestion
        First comment line in an Ion struct is now indented properly.
```

belloale replied:
Fixed.

zslayton commented on resources/plugin-changenotes.html:
```diff
@@ -1,6 +1,33 @@
+        First comment line in an IonStruct is now indented properly.
+    </li>
+    <li>
+        First IonStruct pair after a comment is now indented properly.
```
```suggestion
        First struct field name/value pair after a comment is now indented properly.
```

zslayton commented on resources/plugin-changenotes.html:
```diff
@@ -1,6 +1,33 @@
+        First IonStruct pair after a comment is now indented properly.
+    </li>
+    <li>
+        Space before and after colon in an IonStruct pair has been changed to only a space after the colon to
+        closer match examples in the Ion Spec.
```
```suggestion
        Space before and after the colon in a struct field name/value pair has been changed to only a space
        after the colon to closer match examples in the Ion spec.
```

belloale replied:
Fixed.

zslayton commented on resources/plugin-changenotes.html:
```diff
@@ -1,6 +1,33 @@
+        closer match examples in the Ion Spec.
+    </li>
+    <li>
+        Ion Expressions will be formatted in a single line if its considered a "short" expression, currently a single element.
```
```suggestion
        Ion s-expressions will be formatted in a single line if its considered a "short" s-expression, currently a single element.
```

belloale replied:
Fixed.

zslayton commented on resources/plugin-changenotes.html:
```diff
@@ -1,6 +1,33 @@
+        Ion Expressions will be formatted in a single line if its considered a "short" expression, currently a single element.
+    </li>
+    <li>
+        Comments after the first element in an Ion Expression will remain on the same line as the element.
```
```suggestion
        Comments after the first element in an Ion s-expression will remain on the same line as the element.
```

zslayton reviewed (commented):
The release notes currently refer to implementation details that are likely to be unfamiliar to folks using the plugin. I've added some suggestions that replace these phrases with vanilla Ion concepts. e.g:
* "IonStruct pair" -> "Ion struct field name/value pair"
* "Ion Expression" -> "Ion s-expression"
etc

belloale commented:
@zslayton the latest commit has all the changes recommended. Let me know if you are good for approval. Thanks!

zslayton reviewed (approved):
Looks good, thanks!
